extmod/modusocket.c

```diff
@@ -228,6 +239,42 @@ STATIC mp_obj_t socket_send(mp_obj_t self_in, mp_obj_t buf_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_2(socket_send_obj, socket_send);
 
+STATIC mp_obj_t socket_sendall(mp_obj_t self_in, mp_obj_t buf_in) {
+    mod_network_socket_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    if (self->nic == MP_OBJ_NULL) {
+        // not connected
+        mp_raise_OSError(MP_EPIPE);
+    }
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(buf_in, &bufinfo, MP_BUFFER_READ);
+
+    int _errno;
+    mp_uint_t ret = 0;
+    #if MICROPY_PY_USOCKET_EXTENDED_STATE
+    if (self->timeout == 0) {
+        ret = self->nic_type->send(self, bufinfo.buf, bufinfo.len, &_errno);
+        if (ret == -1) {
+            mp_raise_OSError(_errno);
+        } else if (bufinfo.len > ret) {
+            mp_raise_OSError(MP_EAGAIN);
+        }
+    }
+    #endif
+    // TODO: In CPython3.5, socket timeout should apply to the
+    // entire sendall() operation, not to individual send() chunks.
+    while (bufinfo.len != 0) {
```

Comment:
It looks like this will send out the packet twice, if `MICROPY_PY_USOCKET_EXTENDED_STATE` is enabled and the first call to `->send()` above returns `bufinfo.len`.

Reply:
Yes I think there's a missing `else`, fixed, thanks!